python/cudf/cudf/core/index.py

```diff
@@ -2744,6 +2748,212 @@ def categories(self):
         return self._values.cat().categories
 
 
+def interval_range(
+    start: int = None,
+    end: int = None,
+    periods: int = None,
+    freq: int = None,
+    closed: str = "right",
+    name: str = None,
+) -> "IntervalIndex":
+    """
+    Returns a fixed frequency IntervalIndex.
+
+    Parameters
+    ----------
+    start : numeric, default None
+        Left bound for generating intervals.
+    end : numeric , default None
+        Right bound for generating intervals.
+    periods : int, default None
+        Number of periods to generate
+    freq : numeric, default None
+        The length of each interval. Must be consistent
+        with the type of start and end
+    name : str, default None
+        Name of the resulting IntervalIndex.
+    closed : {"left", "right", "both", "neither"}, default "right"
+        Whether the intervals are closed on the left-side, right-side,
+        both or neither.
+
+    Returns
+    -------
+    IntervalIndex
+
+    Examples
+    --------
+    >>> import cudf
+    >>> import pandas as pd
+    >>> cudf.interval_range(start=0,end=5)
+    IntervalIndex([(0, 0], (1, 1], (2, 2], (3, 3], (4, 4], (5, 5]],
+    ...closed='right',dtype='interval')
+
+    >>> cudf.interval_range(start=0,end=10, freq=2,closed='left')
+    IntervalIndex([[0, 2), [2, 4), [4, 6), [6, 8), [8, 10)],
+    ...closed='left',dtype='interval')
+
+    >>> cudf.interval_range(start=0,end=10, periods=2,closed='left')
+    ...IntervalIndex([[0.0, 3.3333333333333335),
+            [3.3333333333333335, 6.666666666666667),
+            [6.666666666666667, 10.0)],
+            closed='left',
+            dtype='interval')
+
+    """
+    if freq and periods and start and end:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and "
+            "freq, exactly three must be specified"
+        )
+    elif periods and not freq:
+        assert end is not None and start is not None
+        end = end + 1
+        periods_array = cupy.asarray(cupy.arange(start, end))
+        _, bin_edges = cupy.histogram(periods_array, periods)
```

Comment:
Using a histogram here feels a bit overkill. Given we have the number of periods, start, and stop here, we can calculate the step size as `(stop - start) / periods` and then feed that into an `arange` call to generate the intervals.

Reply:
This is working alright-ish, but it looks like it skips out the last bin_edge. 
For example using this code 

```
bin_step = (end-start)/periods
bin_edges = arange(float(start), float(end), float(bin_step))
```
and trying `cudf.interval_range(start=0,end=10, periods=2,closed='left')` the outcome from arange should be 
`array([ 0.,  5. , 10.])` instead I get 
`array([0. , 5.0])`. 

Let me know if there's anything I'm missing that could rectify this :D 

Another approach I think could work if we don't want to use `hist` is 
```
      mn, mx = (periods_array.min(), periods_array.max())
      bin_edges = cupy.linspace(mn, mx, periods + 1, endpoint=True)
```
This gives the correct result, but I'm not sure how much better it would be than `hist`.

Reply:
I would think `linspace` would be more efficient than `histogram`, but not positive and would need benchmarking to confirm.

Under the covers of `arange` we have this libcudf function: https://github.com/rapidsai/cudf/blob/branch-0.19/python/cudf/cudf/_lib/filling.pyx#L102-L115

This seems to cover exactly what you're looking for where it takes in a start, step (freq), and size (periods). We'd need to do a bit of work where if someone only provides an `end` we'd need to translate that into a `start`.

Maybe we should plumb that up to use here?

Reply:
Sounds good! Giving this a try!

Reply:
Yaaay! This works exactly right 🎉